# ***Simple_Hill_Climbing***

---



---



In [1]:
import random
# Distance matrix representing distances between cities
# Replace this with the actual distance matrix for your problem
distance_matrix = [
    [0, 10, 15, 20],
    [10, 0, 35, 25],
    [15, 35, 0, 30],
    [20, 25, 30, 0]
]
def total_distance(path):
    # Calculate the total distance traveled in the given path
    total = 0
    for i in range(len(path) - 1):
        total += distance_matrix[path[i]][path[i+1]]
    total += distance_matrix[path[-1]][path[0]]  # Return to starting city
    return total
def hill_climbing_tsp(num_cities, max_iterations=10000):
    current_path = list(range(num_cities))  # Initial solution, visiting cities in order
    current_distance = total_distance(current_path)
    for _ in range(max_iterations):
        # Generate a neighboring solution by swapping two random cities
        neighbor_path = current_path.copy()
        i, j = random.sample(range(num_cities), 2)
        neighbor_path[i], neighbor_path[j] = neighbor_path[j], neighbor_path[i]
        neighbor_distance = total_distance(neighbor_path)

        # If the neighbor solution is better, move to it
        if neighbor_distance < current_distance:
            current_path = neighbor_path
            current_distance = neighbor_distance
    return current_path
def main():
    num_cities = 4  # Number of cities in the TSP
    solution = hill_climbing_tsp(num_cities)
    print("Optimal path:", solution)
    print("Total distance:", total_distance(solution))
if __name__ == "__main__":
    main()

Optimal path: [0, 1, 3, 2]
Total distance: 80


1. What is distance_matrix?
distance_matrix = [
    [0, 10, 15, 20],
    [10, 0, 35, 25],
    [15, 35, 0, 30],
    [20, 25, 30, 0]
]

It stores distances between cities.

Example:

Distance from city 0 to city 1 = 10
Distance from city 2 to city 3 = 30
2. Why are there zeros on the diagonal?
[0, 10, 15, 20]

Because distance from a city to itself is 0.

Example:

City 0 → 0 = 0
3. What does total_distance(path) do?
def total_distance(path):

This function calculates the total travel cost of a path.

Example:

[0, 1, 3, 2]

means:

0 → 1 → 3 → 2 → back to 0
4. What does this loop do?
for i in range(len(path) - 1):

It moves through the path one city at a time.

If path is:

[0, 1, 3, 2]

then loop checks:

0 → 1
1 → 3
3 → 2
5. Why do we add this line?
total += distance_matrix[path[-1]][path[0]]

Because TSP requires returning to the starting city.

Without it:

0 → 1 → 3 → 2

With it:

0 → 1 → 3 → 2 → 0
6. What does this create?
current_path = list(range(num_cities))

Initial solution.

If num_cities = 4:

[0, 1, 2, 3]

Meaning:

Visit cities in order
7. Why do we use random swapping?
i, j = random.sample(range(num_cities), 2)

To create a neighboring solution.

Example:

Before swap:

[0, 1, 2, 3]

After swapping city 1 and 3:

[0, 3, 2, 1]

Hill climbing explores new routes this way.

8. What does .copy() do?
neighbor_path = current_path.copy()

Creates a separate duplicate list.

Without .copy(), changing neighbor would also change original path accidentally.

9. Why this condition?
if neighbor_distance < current_distance:

Because smaller distance is better in TSP.

If new route costs less:

accept it

Otherwise:

ignore it

This is the “hill climbing” decision.

10. What is the main idea of this algorithm?
Start with one route
↓
Make small random change
↓
Check if route improved
↓
Keep better route
↓
Repeat many times

In [5]:
import numpy as np

# Objective function
def objective(x):
    return -x[0] ** 2 + 5

# Generate neighboring solutions
def generate_neighbors(x, step_size=0.1):
    return [
        np.array([x[0] + step_size]),
        np.array([x[0] - step_size])
    ]

# Hill Climbing Algorithm
def hill_climbing(objective, initial, n_iterations=100, step_size=0.1):

    current = np.array([initial])
    current_eval = objective(current)

    for i in range(n_iterations):

        # Generate neighbors
        neighbors = generate_neighbors(current, step_size)

        # Evaluate neighbors
        neighbor_evals = [objective(n) for n in neighbors]

        # Find best neighbor
        best_idx = np.argmax(neighbor_evals)

        # Move if better solution exists
        if neighbor_evals[best_idx] > current_eval:
            current = neighbors[best_idx]
            current_eval = neighbor_evals[best_idx]

            print(
                f"Step {i+1}: x = {current[0]:.4f}, "
                f"f(x) = {current_eval:.4f}"
            )

        else:
            print("No better neighbors found. Algorithm converged.")
            break

    return current, current_eval


# Initial guess
initial_guess = 2.0

# Run Hill Climbing
solution, value = hill_climbing(
    objective,
    initial_guess,
    n_iterations=100,
    step_size=0.1
)

# Final result
print(f"\nBest solution x = {solution[0]:.4f}, f(x) = {value:.4f}")

Step 1: x = 1.9000, f(x) = 1.3900
Step 2: x = 1.8000, f(x) = 1.7600
Step 3: x = 1.7000, f(x) = 2.1100
Step 4: x = 1.6000, f(x) = 2.4400
Step 5: x = 1.5000, f(x) = 2.7500
Step 6: x = 1.4000, f(x) = 3.0400
Step 7: x = 1.3000, f(x) = 3.3100
Step 8: x = 1.2000, f(x) = 3.5600
Step 9: x = 1.1000, f(x) = 3.7900
Step 10: x = 1.0000, f(x) = 4.0000
Step 11: x = 0.9000, f(x) = 4.1900
Step 12: x = 0.8000, f(x) = 4.3600
Step 13: x = 0.7000, f(x) = 4.5100
Step 14: x = 0.6000, f(x) = 4.6400
Step 15: x = 0.5000, f(x) = 4.7500
Step 16: x = 0.4000, f(x) = 4.8400
Step 17: x = 0.3000, f(x) = 4.9100
Step 18: x = 0.2000, f(x) = 4.9600
Step 19: x = 0.1000, f(x) = 4.9900
Step 20: x = -0.0000, f(x) = 5.0000
No better neighbors found. Algorithm converged.

Best solution x = -0.0000, f(x) = 5.0000


1. What is the purpose of this program?

This program uses the Hill Climbing algorithm to find the maximum value of a mathematical function.

Function:

f(x)=−x
2
+5

The algorithm searches for the best x value that gives the highest output.

2. Why is this called “Hill Climbing”?

Because the algorithm behaves like a person climbing a hill:

It checks nearby positions
Moves toward a higher point
Stops when no nearby point is higher

It always tries to improve step-by-step.

3. What does this function do?
def objective(x):
    return -x[0] ** 2 + 5

This is the function the algorithm wants to optimize.

It receives an input x and returns a score/value.

Example:

x = 0 → value = 5
x = 2 → value = 1

The goal is finding the highest value.

4. Why do we use x[0] instead of just x?

Because x is stored as a NumPy array.

Example:

np.array([2.0])

So:

x[0]

means:

first value inside array
5. What shape does this function have?

The function:

y=−x
2
+5

forms an upside-down parabola.

Highest point at x = 0
Lower values as we move away from 0
6. Why is the maximum at x = 0?

Because:

f(0)=5

which is larger than nearby values:

f(1)=4
f(2)=1

So 0 gives the highest output.

7. What does generate_neighbors() do?
def generate_neighbors(x, step_size=0.1):

It creates nearby possible solutions.

If current value is:

x = 2.0

neighbors become:

2.1 and 1.9

Hill climbing compares these neighbors.

8. Why are neighbors important?

Hill climbing only looks at nearby positions.

It does not examine the whole search space.

Conceptually:

current point
↓
check nearby points
↓
move to better one
9. What is step_size?
step_size = 0.1

This controls how far the algorithm moves each step.

Small step:

slow but precise

Large step:

faster but may skip good solutions
10. What does this line do?
current = np.array([initial])

Stores the starting point as a NumPy array.

If:

initial = 2.0

then:

current = [2.0]
11. Why do we need a starting point?

Hill climbing cannot start from nowhere.

It always needs an initial state.

Example:

start at x = 2

then begin searching nearby.

12. What does this line calculate?
current_eval = objective(current)

It evaluates how good the current solution is.

Example:

If x = 2:

f(2)=−2
2
+5=1

So current score becomes 1.

13. Why do we use a loop?
for i in range(n_iterations):

Hill climbing improves the solution repeatedly.

Each iteration:

Generate neighbors
Compare them
Move if better
14. What happens inside neighbor evaluation?
neighbor_evals = [objective(n) for n in neighbors]

The algorithm calculates scores for all neighbors.

Example:

Current:

x = 2

Neighbors:

2.1 and 1.9

Evaluate both and compare.

15. What does np.argmax() do?
best_idx = np.argmax(neighbor_evals)

It finds the index of the highest value.

Example:

[0.59, 1.39]

Largest value is 1.39.

So algorithm chooses that neighbor.

16. Why this condition?
if neighbor_evals[best_idx] > current_eval:

Hill climbing only moves if the new solution is better.

If not better:

stay where you are
17. What happens when algorithm moves?
current = neighbors[best_idx]

The better neighbor becomes the new current state.

Conceptually:

old position → better position
18. Why does the algorithm stop?
No better neighbors found

This means nearby solutions are worse.

So the algorithm assumes it reached a peak.

For this function, it stops near:

x=0

19. What is “converged” here?

Converged means:

algorithm cannot improve further

No neighboring state gives a higher value.

So searching ends.

20. What is the complete flow of this Hill Climbing algorithm?
Start from initial point
↓
Generate nearby neighbors
↓
Evaluate neighbors
↓
Choose best neighbor
↓
If better → move there
↓
Repeat
↓
Stop when no improvement exists